File: BCW3/C++/Pen&Paper/Paper.cpp
#include <iostream>

#include "Paper.h"
#include "Pen.h"

Paper::Paper(int maxSymbols) {
    this->maxSymbols = maxSymbols;
    this->symbols = 0;
    this->content = "";
}

Paper::~Paper() {}

int Paper::getMaxSymbols() const {
    return this->maxSymbols;
}

int Paper::getSymbols() const {
    return this->symbols;
}

void Paper::addContent(const std::string& message) {
    int freeSpace = this->maxSymbols - this->symbols;
    
    if ( this->symbols == this->maxSymbols ) {
        throw OutOfSpace();
    }
    
    this->content += message.substr(0, freeSpace);
    
    this->symbols += this->content.size();
}

void Paper::show() const {
    std::cout << this->content << std::endl;
}

File: BCW3/java/Pen&Paper/Paper.java
public class Paper {
    private StringBuffer content;
    private int maxSymbols;
    private int symbols;

    public Paper(int maxSymbols) {
        this.maxSymbols = maxSymbols;
        content = new StringBuffer();
        symbols = 0;
    }

    public Paper() {
        this(4096);
    }

    public int getMaxSymbols() {
        return maxSymbols;
    }
    
    public int getSymbols() {
        return symbols;
    }

    public void addContent(String message) throws OutOfSpaceException {
        int total = content.length() + message.length();

        if ( content.length() == maxSymbols ) {
            throw new OutOfSpaceException();
        }
        if ( total > maxSymbols ) {
            content.append(message.substring(0, maxSymbols-content.length()));
            return;
        }
        content.append(message);
        symbols += content.length();
    }
    public void show() {
        System.out.println(content);
    }
}